src/func/Microsoft.Health.Fhir.Ingest.Host/IomtConnectorFunctions.cs

```diff
@@ -30,18 +27,17 @@ public IomtConnectorFunctions(ITelemetryLogger logger)
         }
 
         [FunctionName("MeasurementCollectionToFhir")]
-        public async Task<IActionResult> MeasurementCollectionToFhir(
-            [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpRequest req,
+        public async Task MeasurementCollectionToFhir(
+            [Common.EventHubs.EventHubTrigger("fhir", Connection = "OutputEventHub")] EventData[] events,
             [Blob("template/%Template:FhirMapping%", FileAccess.Read)] string templateDefinition,
             [MeasurementFhirImport] MeasurementFhirImportService measurementImportService)
         {
             EnsureArg.IsNotNull(measurementImportService, nameof(measurementImportService));
-            EnsureArg.IsNotNull(req, nameof(req));
+            EnsureArg.IsNotNull(events, nameof(events));
 
             try
             {
-                await measurementImportService.ProcessStreamAsync(req.Body, templateDefinition, _logger).ConfigureAwait(false);
-                return new AcceptedResult();
+                await measurementImportService.ProcessEventsAync(events, templateDefinition, _logger).ConfigureAwait(false);

```

Comment:
This may just be an edge case, but am wondering there is any state that was essentially managed in stream analytics previously that we would need to now manage in the azure function app.

For example if the function app was restarted or shut down, do we know exactly where to pick up again? I am also curious what determines whether the checkpoint advances (does it advance when we begin processing an event, or when we successfully complete processing an event).

Reply:
Great question.  This is all managed in the Event Hub trigger, specifically the EventHubListener class.  Your observation is precisely why we can't just buffer in the function itself.  If we collect events in the Function the watermark will still progress as if we successfully processed them and those events would be lost if there is an error, restart, deployment, etc.

The watermark is saved as part of the CheckpointAsync call at the end of the ConsumeEvents method which is after a batch has been created and sent to the Function for processing.  An interruption prior to the watermark update would just cause the system to resume from the last known watermark and start over again.  Event Hub architectures are "at least once" delivery where you are guaranteed you will see the message once but multiple times is possible.  Our processing downstream is designed to handle this so data isn't duplicated.